Firestore/core/src/firebase/firestore/util/hashing.h

```diff
@@ -182,6 +183,22 @@ auto RankedInvokeHash(const Range& range, HashChoice<3>)
   return result;
 }
 
+/**
+ * Hashes the contents of the given optional value, only if the underlying
+ * value can itself be hashed.
+ */
+template <typename K>
+auto RankedInvokeHash(const absl::optional<K>& option, HashChoice<4>)
+    -> decltype(InvokeHash(*option)) {
+  return option ? InvokeHash(*option) : 0;
```

Comment:
This makes `absl::make_optional<int>(0)` has to the same value as `absl::nullopt`, though it's probably no big deal?

Reply:
It's not much of an issue. An alternative would be something like this?
```
  size_t result = 1;
  result = Combine(result, InvokeHash(option.has_value()));
  result = Combine(result, option ? InvokeHash(*option) : 0);
  return result;
```

However, all this does is change it such that `Hash(absl::nullopt)` collides with `Hash(absl::optional<int>(-31))`.

I suppose that's arguably a better outcome because -31 is not anywhere near as common as 0. WDYT?